File: src/main/java/de/wittenbude/exportify/models/Artist.java
package de.wittenbude.exportify.models;

import de.wittenbude.exportify.models.embeds.Followers;
import de.wittenbude.exportify.models.embeds.Image;
import de.wittenbude.exportify.models.embeds.ObjectType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@Getter
@Setter
@Accessors(chain = true)
@Entity
public class Artist {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, String> externalUrls;

    @JdbcTypeCode(SqlTypes.JSON)
    private Followers followers;

    @JdbcTypeCode(SqlTypes.JSON)
    private List<String> genres;

    private String href;

    @Column(unique = true)
    private String spotifyID;

    @JdbcTypeCode(SqlTypes.JSON)
    private List<Image> images;

    private String name;

    private Integer popularity;

    private ObjectType objectType;

    private String uri;
}

File: src/main/kotlin/de/wittenbude/exportify/models/Artist.kt
@file:Suppress("PLUGIN_IS_NOT_ENABLED")

package de.wittenbude.exportify.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import se.michaelthelin.spotify.enums.ModelObjectType
import se.michaelthelin.spotify.model_objects.specification.Artist as ApiArtist

@Serializable
data class Artist(
    val externalUrls: Map<String, String>? = null,
    val followerHref: String? = null,
    val followerTotal: Int? = null,
    val genres: List<String>? = null,
    val href: String? = null,
    @SerialName("_id") val id: String,
    val images: List<String>? = null,
    val name: String,
    val popularity: Int? = null,
    val type: ModelObjectType? = null,
    val uri: String? = null,
)


fun ApiArtist.convert() : Artist {
    return Artist(
        externalUrls = externalUrls?.externalUrls,
        followerHref = followers?.href,
        followerTotal = followers?.total,
        genres = genres?.asList(),
        href = href,
        id = id,
        images = images?.map { it.url },
        name = name,
        popularity = popularity,
        type = type,
        uri = uri
    )
}